app/src/main/kotlin/org/equeim/tremotesf/ui/addtorrent/AddTorrentFileFragment.kt

```diff
@@ -75,20 +75,30 @@ class AddTorrentFileFragment : AddTorrentFragment(
 
             if (savedInstanceState == null) {
                 fragment.lifecycleScope.launch {
-                    downloadDirectoryEdit.setText(
-                        (if (Settings.rememberDownloadDirectory.get()) {
-                            GlobalServers.serversState.value.currentServer?.lastDownloadDirectory?.takeIf { it.isNotEmpty() }
-                        } else {
-                            null
-                        } ?: GlobalRpc.serverSettings.downloadDirectory).toNativeSeparators()
-                    )
+                    val downloadDirectory = if (Settings.rememberDownloadDirectory.get()) {
+                        GlobalServers.serversState.value.currentServer?.lastDownloadDirectory?.takeIf { it.isNotEmpty() }
+                    } else {
+                        null
+                    }
+
+                    if (downloadDirectory != null) {
+                        downloadDirectoryEdit.setText(downloadDirectory.toNativeSeparators())
+                    } else {
+                        GlobalRpc.isConnected.launchAndCollectWhenStarted(fragment.viewLifecycleOwner) {
+                            downloadDirectoryEdit.setText(GlobalRpc.serverSettings.downloadDirectory.toNativeSeparators())
+                        }
+                    }
                 }
             }
 
             val directoriesAdapter =
                 AddTorrentDirectoriesAdapter(downloadDirectoryEdit, savedInstanceState)
             downloadDirectoryEdit.setAdapter(directoriesAdapter)
 
+            GlobalRpc.isConnected.launchAndCollectWhenStarted(fragment.viewLifecycleOwner) {
+                directoriesAdapter.refresh()
```

Comment:
This means that we will immediately lose restored state when refresh() is called, meaning that AddTorrentDirectoriesAdapter.remove() calls (when entries are removed from the list by user) will be lost.

I think that `directories()` should be called only if we haven't restored state, as before. And directories would be a suspending function that simply waits until GlobalRpc.isConnected == true (you can pass fragment.viewLifecycleOwner.lifecycleScope to AddTorrentDirectoriesAdapter constructor and launch coroutine from there).

Reply:
Something like this? (I just force pushed)